src/qibolab/_core/instruments/qrng.py

```diff
@@ -0,0 +1,151 @@
+from typing import Iterable, Optional, Union
+
+import numpy as np
+import numpy.typing as npt
+from scipy.linalg import toeplitz
+from serial import Serial
+
+from .abstract import Instrument
+
+__all__ = ["QRNG"]
+
+
+def unpackbits(x: npt.NDArray, num_bits: int) -> npt.NDArray:
+    if np.issubdtype(x.dtype, np.floating):
+        raise ValueError("numpy data type needs to be int-like")
+    xshape = list(x.shape)
+    x = x.reshape([-1, 1])
+    mask = 2 ** np.arange(num_bits, dtype=x.dtype).reshape([1, num_bits])
+    return (x & mask).astype(bool).astype(int).reshape(xshape + [num_bits])
+
+
+def generate_toeplitz(input_bits: int, extraction_ratio: int) -> npt.NDArray:
+    """Generate a pseudo-random Toeplitz matrix of dimension ``(input_bits, extraction_ratio)``."""
+    while True:
+        c = np.mod(np.random.permutation(input_bits), 2)
+        r = np.mod(np.random.permutation(extraction_ratio), 2)
```

Comment:
This is the part I understand the least: why do you need pseudo-random numbers to generate random numbers?

Reply:
> Do you have a reference for the algorithm?

Some references from the original author are https://arxiv.org/pdf/2402.09481 appendix A5 and https://github.com/CQCL/cryptomite, but I did not check in detail yet.

The outcome summarized is that the device produces 12-bit integers following a Gaussian distribution (these are accessible via `qrng.read()`) and the extraction procedure converts them to 4-bit integers following a uniform distribution (accessible via `qrng.extract()`). Then I concatenate multiple 4-bit samples to go to 32-bits (floats) (`qrng.random()`), which btw I am not sure if it is correct.